src/cmd/main.cpp

```diff
@@ -416,6 +412,26 @@ void search(const Options& opts) {
     callback = &nullWriter;
     hinfo.reset(new HitCounterInfo);
   }
+  else if (opts.BeforeContext > -1 || opts.AfterContext > -1) {
+    if (opts.PrintPath) {
```

Comment:
This block of code is annoying since the only thing that differs is the type, so there are 5-6 duplicated LoC.

Reply:
One could do it like this in C++14:

    auto cfunc = [&](auto&&... args) {
      return opts.PrintPath ?
        new LineContextPathWriterInfo(args...) :
        new LineContextHitWriterInfo(args...);
    };  

    callback = opts.PrintPath ? &lineContextPathWriter : &lineContextHitWriter;

    hinfo.reset(cfunc(
      opts.openOutput(), pmap.get(),
      std::max(opts.BeforeContext, 0),
      std::max(opts.AfterContext, 0),
      opts.GroupSeparator
    ));

It's not really shorter, though.